heartbeat/IPSEC

```diff
@@ -0,0 +1,163 @@
+. ${OCF_FUNCTIONS_DIR}/ocf-shellfuncs
+
+#######################################################################
+
+meta_data() {
+	cat <<END
+<?xml version="1.0"?>
+<!DOCTYPE resource-agent SYSTEM "ra-api-1.dtd">
+<resource-agent name="ipsec">
+<version>1.0</version>
+
+<longdesc lang="en">
+This is a Resource Agent to manage IPSEC tunnels associated with a VIP.
+It's meant to be collocated with a specific VIP, and will manage
+setting up or down a specific tunnel.
+</longdesc>
+<shortdesc lang="en">Handles IPSEC tunnels for VIPs</shortdesc>
+
+<parameters>
+<parameter name="tunnel" unique="1" required="1">
+<longdesc lang="en">
+The name of the tunnel to be monitored.
+</longdesc>
+<shortdesc lang="en">Tunnel name</shortdesc>
+<content type="string" default="" />
+</parameter>
+<parameter name="vip" unique="1" required="1">
+<longdesc lang="en">
+VIP that the tunnel is using.
+</longdesc>
+<shortdesc lang="en">VIP</shortdesc>
+<content type="string" default="" />
+</parameter>
+</parameters>
+
+<actions>
+<action name="start"        timeout="20" />
+<action name="stop"         timeout="20" />
+<action name="monitor"      timeout="20" interval="10" depth="0" />
+<action name="reload"       timeout="20" />
+<action name="meta-data"    timeout="5" />
+</actions>
+</resource-agent>
+END
+}
+
+#######################################################################
+
+ipsec_usage() {
+	cat <<END
+usage: $0 {start|stop|monitor|validate-all|meta-data}
+
+Expects to have a fully populated OCF RA-compliant environment set. And
+should have a collocation constraint with a VIP associated with the
+tunnel.
+END
+}
+
+ipsec_start() {
+	ipsec auto --add "${OCF_RESKEY_tunnel}"
+	ipsec whack --listen &>> /tmp/ipsec-agent.log
+	local return_code=$?
+	if [ $return_code -eq  1 -o $return_code -eq 10 ]; then
+		ocf_log warn "${OCF_RESOURCE_INSTANCE} : Unable to add tunnel ${OCF_RESKEY_tunnel} with return code ${return_code}"
+		return $OCF_ERR_GENERIC
+	else
+		return $OCF_SUCCESS
+	fi
+}
+
+ipsec_stop() {
+	ipsec auto --down "${OCF_RESKEY_tunnel}"
+	local return_code=$?
+	ocf_log info "${OCF_RESOURCE_INSTANCE} : Put down tunnel ${OCF_RESKEY_tunnel} with return code ${return_code}"
+	return $OCF_SUCCESS
+}
+
+ipsec_monitor() {
+	# Monitor _MUST!_ differentiate correctly between running
+	# (SUCCESS), failed (ERROR) or _cleanly_ stopped (NOT RUNNING).
+	# That is THREE states, not just yes/no.
+
+	ipsec status | grep "$OCF_RESKEY_tunnel" | grep -q unoriented
+	state=$?
+	if [ "$state" == "0" ]; then
+		ip addr show | grep -q "${OCF_RESKEY_vip}"
+		hosting_vip=$?
+		if [ "hosting_vip" == "0" ]; then
+			ocf_log warn "${OCF_RESOURCE_INSTANCE} : tunnel ${OCF_RESKEY_tunnel} is unoriented"
+			return $OCF_ERR_GENERIC
+		else
+			return $OCF_NOT_RUNNING
+		fi
+	else
+		return $OCF_SUCCESS
+	fi
+}
+
+ipsec_validate() {
+	# The tunnel needs to be defined in the configuration
+	cat /etc/ipsec.d/*.conf | grep -q "conn $OCF_RESKEY_tunnel"
+	state=$?
```

Comment:
Maybe it'd be better to use a conf-parameter to specify the .conf-file and do 
"if [ -f ... ]" instead, and maybe a confdir parameter that defaults to "/etc/ipsec.d/" that could later be part of e.g. configure auto-detection based on distribution/OS.

Reply:
I could do "/etc/ipsec.d/" in a confdir parameter. But do we really need to specify the parameter? I dynamically generate the names, where I use this in my ansible role, I dynamically generate the configurations (and their names), so I don't fully know which configuration contains which tunnel. I could just stick them all in one file, but then it gets harder to debug later. This just makes life way simpler :/

Reply:
Ah. In that case I guess this is OK, so the confdir parameter would be nice to support all distros/OSs and other scenarios.